pyro/contrib/funsor/infer/traceenum_elbo.py

```diff
@@ -148,19 +148,33 @@ def differentiable_loss(self, model, guide, *args, **kwargs):
             costs = contracted_costs + uncontracted_factors  # model costs: logp
             costs += [-f for f in guide_terms["log_factors"]]  # guide costs: -logq
 
-            # finally, integrate out guide variables in the elbo and all plates
             plate_vars = guide_terms["plate_vars"] | model_terms["plate_vars"]
-            elbo = to_funsor(0, output=funsor.Real)
+            # compute the marginal logq in the guide corresponding to each cost term
+            targets = dict()
             for cost in costs:
-                # compute the marginal logq in the guide corresponding to this cost term
-                log_prob = funsor.sum_product.sum_product(
+                input_vars = frozenset(cost.inputs.keys())
+                if input_vars not in targets:
+                    targets[input_vars] = funsor.Tensor(
+                        funsor.ops.new_zeros(
+                            funsor.tensor.get_default_prototype(), ()
```

Comment:
Use `cost` as a prototype here

Reply:
`cost` can be a lazy expression in which case i'm not sure how to retrieve its prototype.